Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt

```diff
@@ -0,0 +1,670 @@
+        val prefix = when (paramStub.qualifier) {
+            is ConstructorParamStub.Qualifier.VAL -> if (paramStub.qualifier.overrides) "override val " else "val "
+            is ConstructorParamStub.Qualifier.VAR -> if (paramStub.qualifier.overrides) "override var " else "var "
+            ConstructorParamStub.Qualifier.NONE -> ""
+        }
+        return "$prefix${paramStub.name}: ${renderStubType(paramStub.type)}"
+    }
+
+    private fun renderSuperInit(superClassInit: SuperClassInit): String {
+        val parameters = superClassInit.arguments.joinToString(prefix = "(", postfix = ")") { renderValueUsage(it) }
+        return "${renderStubType(superClassInit.type)}$parameters"
+    }
+
+    private fun renderStubType(stubType: StubType): String = when (stubType) {
+        is WrapperStubType -> stubType.kotlinType.render(kotlinFile)
+        is ClassifierStubType -> kotlinFile.reference(stubType.classifier) + renderTypeArguments(stubType.typeArguments)
+        is SymbolicStubType -> stubType.name + if (stubType.nullable) "?" else ""
+    }
+
+    private fun renderValueUsage(value: ValueStub): String = when (value) {
+        is StringConstantStub -> value.value
+        is IntegralConstantStub -> renderIntegralConstant(value)!!
+        is DoubleConstantStub -> renderDoubleConstant(value)!!
+        is ConstructorParamStub -> value.name
+    }
+
+    private fun renderAnnotation(annotationStub: AnnotationStub): String = when (annotationStub) {
+        AnnotationStub.ObjC.ConsumesReceiver -> "@CCall.ConsumesReceiver"
+        AnnotationStub.ObjC.ReturnsRetained -> "@CCall.ReturnsRetained"
+        is AnnotationStub.ObjC.Method -> {
+            val stret = if (annotationStub.isStret) ", true" else ""
+            val selector = annotationStub.selector.quoteAsKotlinLiteral()
+            val encoding = annotationStub.encoding.quoteAsKotlinLiteral()
+            "@ObjCMethod($selector, $encoding$stret)"
+        }
+        is AnnotationStub.ObjC.Factory -> {
+            val stret = if (annotationStub.isStret) ", true" else ""
+            val selector = annotationStub.selector.quoteAsKotlinLiteral()
+            val encoding = annotationStub.encoding.quoteAsKotlinLiteral()
+            "@ObjCFactory($selector, $encoding$stret)"
+        }
+        AnnotationStub.ObjC.Consumed ->
+            "@CCall.Consumed"
+        is AnnotationStub.ObjC.Constructor ->
+            "@ObjCConstructor(${annotationStub.selector.quoteAsKotlinLiteral()}, ${annotationStub.designated})"
+        is AnnotationStub.ObjC.ExternalClass -> {
+            val protocolGetter = annotationStub.protocolGetter.quoteAsKotlinLiteral()
+            val binaryName = annotationStub.binaryName.quoteAsKotlinLiteral()
+            "@ExternalObjCClass" + when {
+                annotationStub.protocolGetter.isEmpty() && annotationStub.binaryName.isEmpty() -> ""
+                annotationStub.protocolGetter.isEmpty() -> "($binaryName)"
```

Comment:
Is this correct?

Reply:
It looks correct to me. What do you mean?

Reply:
If you supply only one argument to `@ExternalObjCClass` annotation, it will be bound to its first parameter, `protocolGetter`.

Reply:
Shame on me.